packages/beacon-node/src/chain/shufflingCache.ts

```diff
@@ -0,0 +1,170 @@
+        shuffling = state.epochCtx.currentShuffling;
+        break;
+      case state.epochCtx.previousShuffling.epoch:
+        shuffling = state.epochCtx.previousShuffling;
+        break;
+      default:
+        throw new Error(`Shuffling not found from state ${state.slot} for epoch ${shufflingEpoch}`);
+    }
+
+    let found = false;
+    for (const item of this.items) {
+      if (item.epoch === shufflingEpoch && item.decisionBlockHex === decisionBlockHex) {
+        found = true;
+        if (isPromiseCacheItem(item)) {
+          // unblock consumers of this promise
+          item.resolveFn(shuffling);
+          // then update item type to shuffling
+          Object.assign(item, {type: CacheItemType.shuffling, shuffling});
+          // TODO: remove promise and resolveFn?
+          // we updated type to CacheItemType.shuffling so the above fields are not used anyway
+          this.metrics?.shufflingCache.processStateUpdatePromise.inc();
+        } else {
+          // ShufflingCacheItem, do nothing
+          this.metrics?.shufflingCache.processStateNoOp.inc();
+        }
+        break;
+      }
+    }
+
+    if (!found) {
+      this.add({type: CacheItemType.shuffling, epoch: shufflingEpoch, decisionBlockHex, shuffling});
+      this.metrics?.shufflingCache.processStateInsertNew.inc();
+    }
+  }
+
+  /**
+   * Insert a promise to make sure we don't regen state for the same shuffling.
+   * Bound by MAX_SHUFFLING_PROMISE to make sure our node does not blow up.
+   */
+  insertPromise(shufflingEpoch: Epoch, dependentRootHex: RootHex): void {
+    const promiseCount = this.items.filter((item) => isPromiseCacheItem(item)).length;
+    if (promiseCount >= MAX_SHUFFLING_PROMISE) {
+      throw new Error(
+        `Too many shuffling promises: ${promiseCount}, shufflingEpoch: ${shufflingEpoch}, dependentRootHex: ${dependentRootHex}`
+      );
+    }
+    let resolveFn: ((shuffling: EpochShuffling) => void) | null = null;
+    const promise = new Promise<EpochShuffling>((resolve) => {
+      resolveFn = resolve;
+    });
+    if (resolveFn === null) {
+      throw new Error("Promise Constructor was not executed immediately");
+    }
+
+    const cacheItem: PromiseCacheItem = {
+      type: CacheItemType.promise,
+      epoch: shufflingEpoch,
+      decisionBlockHex: dependentRootHex,
+      promise,
+      resolveFn,
+    };
+    this.add(cacheItem);
+    this.metrics?.shufflingCache.insertPromiseCount.inc();
+  }
+
+  /**
+   * Most of the time, this should return a shuffling immediately.
+   * If there's a promise, it means we are computing the same shuffling, so we wait for the promise to resolve.
+   * Return null if we don't have a shuffling for this epoch and dependentRootHex.
+   */
+  async get(shufflingEpoch: Epoch, dependentRootHex: RootHex): Promise<EpochShuffling | null> {
+    for (const item of this.items) {
+      if (item.epoch === shufflingEpoch && item.decisionBlockHex === dependentRootHex) {
```

Comment:
Why not use a map?

Reply:
this started with an array to implement simple LRU cache, will switch to map of epoch 👍